### PR TITLE
V3 client TLS

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

File: clientv3/client.go [hunk 1]
```diff
@@ -99,22 +103,43 @@ func (c *Client) Errors() (errs []error) {
...
+	}
+	conn, err := grpc.Dial(endpoint, opts...)
 	if err != nil {
 		return nil, err
 	}
 	return conn, nil
 }
 
-func newClient(conn *grpc.ClientConn, cfg *Config) *Client {
+func newClientNoErr(conn *grpc.ClientConn, cfg *Config) *Client {
```
Comment (xiang90) [inline clientv3/client.go]: we usually name this as `mustNewClient`

File: etcdctlv3/main.go [hunk 2]
```diff
@@ -42,6 +42,15 @@ var (
 
 func init() {
 	rootCmd.PersistentFlags().StringVar(&globalFlags.Endpoints, "endpoint", "127.0.0.1:2378", "gRPC endpoint")
+	rootCmd.PersistentFlags().StringVar(
```
Comment (xiang90) [inline etcdctlv3/main.go]: do we really need multiple lines here?

Comment (xiang90) [inline etcdctlv3/main.go]: also how about just cert instead of cert-file and key instead of key-file?

File: integration/cluster.go [hunk 3]
```diff
@@ -80,7 +88,7 @@ func (c *cluster) fillClusterForMembers() error {
 	addrs := make([]string, 0)
 	for _, m := range c.Members {
 		scheme := "http"
-		if !m.PeerTLSInfo.Empty() {
+		if m.PeerTLSInfo != nil {
```
Comment (xiang90) [inline integration/cluster.go]: why we cannot use Empty anymore?

File: integration/v3_grpc_test.go [hunk 4]
```diff
@@ -1416,3 +1417,86 @@ func testLeaseRemoveLeasedKey(t *testing.T, act func(*ClusterV3, int64) error) {
...
+func newClusterV3NoClients(t *testing.T, cfg *ClusterConfig) *ClusterV3 {
+	cfg.UseV3 = true
+	cfg.UseGRPC = true
+	clus := &ClusterV3{cluster: NewClusterByConfig(t, cfg)}
+	clus.Launch(t)
+	return clus
+}
+
+// TestTLSGRPCInsecureClient checks that connection is rejected if server is TLS but not client.
+func TestTLSGRPCInsecureClient(t *testing.T) {
```
Comment (xiang90) [inline integration/v3_grpc_test.go]: TestTLSGRPCRejectInscureClient

File: integration/v3_grpc_test.go [hunk 5]
```diff
@@ -1416,3 +1417,86 @@ func testLeaseRemoveLeasedKey(t *testing.T, act func(*ClusterV3, int64) error) {
...
+	defer testutil.AfterTest(t)
+
+	cfg := ClusterConfig{Size: 3, ClientTLS: &testTLSInfo}
+	clus := newClusterV3NoClients(t, &cfg)
+	defer clus.Terminate(t)
+
+	clus.Members[0].ClientTLSInfo = nil
```
Comment (xiang90) [inline integration/v3_grpc_test.go]: add a comment for why we clear the ClientTLSInfo here?

File: integration/v3_grpc_test.go [hunk 6]
```diff
@@ -1416,3 +1417,86 @@ func testLeaseRemoveLeasedKey(t *testing.T, act func(*ClusterV3, int64) error) {
...
+	ctx, cancel := context.WithTimeout(context.TODO(), 5*time.Second)
+	donec := make(chan error, 1)
+	go func() {
+		reqput := &pb.PutRequest{Key: []byte("foo"), Value: []byte("bar")}
+		_, err := client.KV.Put(ctx, reqput)
+		donec <- err
+	}()
+	conn := client.ActiveConnection()
+	st, _ := conn.State()
+	if _, err := conn.WaitForStateChange(ctx, st); err != nil {
```
Comment (xiang90) [inline integration/v3_grpc_test.go]: shall we check the changed state?

File: integration/v3_grpc_test.go [hunk 7]
```diff
@@ -1416,3 +1417,86 @@ func testLeaseRemoveLeasedKey(t *testing.T, act func(*ClusterV3, int64) error) {
...
+		t.Fatalf("unexpected error waiting for change (%v)", err)
+	}
+	cancel()
+	if perr := <-donec; perr == nil {
+		t.Fatalf("expected client error on put")
+	}
+}
+
+// TestTLSSecureGRPCClient checks that connection is rejected if client is TLS but not server.
+func TestTLSGRPCSecureClient(t *testing.T) {
```
Comment (xiang90) [inline integration/v3_grpc_test.go]: TestInscureGPRCRejectTLSClient?

File: tools/benchmark/cmd/root.go [hunk 8]
```diff
@@ -40,12 +41,25 @@ var (
...
+
 	cpuProfPath string
 	memProfPath string
 )
 
 func init() {
 	RootCmd.PersistentFlags().StringVar(&endpoints, "endpoint", "127.0.0.1:2378", "comma-separated gRPC endpoints")
+	RootCmd.PersistentFlags().StringVar(
+		&tls.CertFile,
+		"cert-file",
```
Comment (xiang90) [inline tools/benchmark/cmd/root.go]: same here as comments for etcdctlv3

File: integration/cluster.go (hunk 3 above)
Comment (heyitsanthony) [inline integration/cluster.go]: I want an empty tls to fail through the regular tls machinery; if a tls is given it should have cert data.

Comment (xiang90) [inline integration/cluster.go]: hrm... OK. I think in the real etcdserver codebase, tls is always created. So nil case should not exist. But I need to double check. It is fine here though.

Comment (heyitsanthony) [inline integration/cluster.go]: Right, there are some places where the nil gets replaced with an empty tls so bits that expect it are happy. The change was primarily meant to sanely rip out the goofy `UsePeerTLS` flag via nil pointers.

**Comment by heyitsanthony** [comment]
all fixed up ptal /cc @xiang90

File: integration/v3_grpc_test.go (hunk 7 above)
Comment (xiang90) [inline integration/v3_grpc_test.go]: @heyitsanthony also this one?

File: etcdctlv3/command/global.go [hunk 9]
```diff
@@ -30,7 +32,22 @@ func mustClient(cmd *cobra.Command) *clientv3.Client {
 	if err != nil {
 		ExitWithError(ExitError, err)
 	}
-	client, err := clientv3.NewFromURL(endpoint)
+
+	// set tls if any one tls option set
+	var cfgtls *transport.TLSInfo
+	tls := transport.TLSInfo{}
+	if tls.CertFile, err = cmd.Flags().GetString("cert-file"); err == nil {
```
Comment (xiang90) [inline etcdctlv3/command/global.go]: cert-file -> cert (curl uses cert)

File: etcdctlv3/command/global.go [hunk 10]
```diff
@@ -30,7 +32,22 @@ func mustClient(cmd *cobra.Command) *clientv3.Client {
...
+		cfgtls = &tls
+	}
+	if tls.KeyFile, err = cmd.Flags().GetString("key-file"); err == nil {
```
Comment (xiang90) [inline etcdctlv3/command/global.go]: key-file > key (curl uses key)

basically we wanted to copy curl flag options.

File: tools/benchmark/cmd/root.go [hunk 11]
```diff
@@ -40,12 +41,16 @@ var (
...
 
+	tls transport.TLSInfo
+
 	cpuProfPath string
 	memProfPath string
 )
 
 func init() {
 	RootCmd.PersistentFlags().StringVar(&endpoints, "endpoint", "127.0.0.1:2378", "comma-separated gRPC endpoints")
+	RootCmd.PersistentFlags().StringVar(&tls.CertFile, "cert-file", "", "identify HTTPS client using this SSL certificate file")
```
Comment (xiang90) [inline tools/benchmark/cmd/root.go]: add key flag too?

File: etcdctlv3/command/global.go (hunk 9 above)
Comment (heyitsanthony) [inline etcdctlv3/command/global.go]: so break from etcdctlv2's flags?

File: etcdctlv3/command/global.go (hunk 10 above)
Comment (heyitsanthony) [inline etcdctlv3/command/global.go]: there shouldn't be a key-file at all, though?

File: tools/benchmark/cmd/root.go (hunk 11 above)
Comment (heyitsanthony) [inline tools/benchmark/cmd/root.go]: you asked me to remove it?

**Comment by heyitsanthony** [comment]
ok flags fixed

**Comment by xiang90** [comment]
LGTM. Thanks!

**Comment by heyitsanthony** [comment]
ok, will rebase all fixup patches then merge on greenlight

**Pull request merged**